metadata-ingestion/src/datahub/testing/golden_diff.py

```diff
@@ -0,0 +1,233 @@
+import re
+from collections import defaultdict
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Literal, Sequence, Set, Tuple, Union
+
+import deepdiff.serialization
+import inflection
+import yaml
+from deepdiff import DeepDiff
+from deepdiff.model import DiffLevel
+
+ReportType = Literal[
+    "type_changes",
+    "dictionary_item_added",
+    "dictionary_item_removed",
+    "values_changed",
+    "unprocessed",
+    "iterable_item_added",
+    "iterable_item_removed",
+    "iterable_item_moved",
+    "attribute_added",
+    "attribute_removed",
+    "set_item_added",
+    "set_item_removed",
+    "repetition_change",
+]
+
+
+@dataclass(frozen=True)
+class GoldenAspect:
+    idx: int  # Location in list of MCEs in golden file
+    urn: str
+    change_type: str
+    aspect_name: str
+    aspect: Dict[str, Any] = field(hash=False)
+    original: Any = field(hash=False)
+
+    @classmethod
+    def create_from_mcp(cls, idx: int, obj: Dict[str, Any]) -> "GoldenAspect":
+        aspect = obj["aspect"]
+        return cls(
+            idx=idx,
+            urn=obj["entityUrn"],
+            change_type=obj["changeType"],
+            aspect_name=obj["aspectName"],
+            aspect=aspect.get("json", aspect),
+            original=obj,
+        )
+
+
+AspectsByUrn = Dict[str, Dict[str, List[GoldenAspect]]]
+
+
+def get_aspects_by_urn(obj: object) -> AspectsByUrn:
+    """Restructure a list of MCPs by urn and aspect.
+    Retains information like the original dict and index to facilitate `apply_delta` later.
+
+    Raises:
+        AssertionError: If the input is not purely a list of MCPs.
+    """
+    d: AspectsByUrn = defaultdict(dict)
+    assert isinstance(obj, list), obj
+    for i, entry in enumerate(obj):
+        assert isinstance(entry, dict), entry
+        if "proposedSnapshot" in entry:
+            raise AssertionError("Found MCEs in output")
+        elif "entityUrn" in entry and "aspectName" in entry and "aspect" in entry:
+            urn = entry["entityUrn"]
+            aspect_name = entry["aspectName"]
+            aspect = GoldenAspect.create_from_mcp(i, entry)
+            d[urn].setdefault(aspect_name, []).append(aspect)
+        else:
+            raise AssertionError(f"Unrecognized MCE: {entry}")
+
+    return d
+
+
+@dataclass
+class AspectDiff:
+    diff: DeepDiff
+    aspects_added: Dict[int, GoldenAspect] = field(init=False, default_factory=dict)
+    aspects_removed: Dict[int, GoldenAspect] = field(init=False, default_factory=dict)
+    aspects_changed: Dict[
+        Tuple[int, GoldenAspect, GoldenAspect], List[DiffLevel]
+    ] = field(init=False, default_factory=lambda: defaultdict(list))
+
+    def __post_init__(self):
```

Comment:
make this a classmethod `create` that then creates the `AspectDiff` obj at the end

Reply:
I had this originally and swapped to post init. Do you not like post init, or is there any style recs we're following here?

Reply:
more a personal style thing

If we're just initializing one small field, using post_init is fine. In this case we're basically initializing the entire thing in post init, which feels like an abuse of the post init mechanism